src/redisearch_rs/slots_tracker/src/slots_tracker.rs

```diff
@@ -0,0 +1,932 @@
+                    Self::Stable(unsafe { NonZeroU32::new_unchecked(1) })
+                }
+            }
+            Self::Unstable => {
+                debug_assert!(false, "Cannot increment Unstable version");
+                Self::Unstable
+            }
+        }
+    }
+}
+
+/// Safe slots tracker implementation.
+///
+/// This structure encapsulates all slot tracking state and provides safe methods
+/// for manipulating the three slot sets and version counter.
+///
+/// All methods take `&mut self`, ensuring compile-time enforcement of exclusive access.
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+pub struct SlotsTracker {
+    /// Local responsibility slots - owned by this Redis instance in the cluster topology.
+    local: SlotSet,
+    /// Fully available non-owned slots - locally available but not owned by this instance.
+    fully_available: SlotSet,
+    /// Partially available non-owned slots - partially available and not owned by this instance.
+    partially_available: SlotSet,
+    /// Version counter for tracking changes to the slots configuration.
+    /// Incremented whenever the slot configuration changes. Wraps around safely using wrapping arithmetic.
+    /// This is always a Stable variant internally; Unstable is only returned by check_availability
+    /// when the configuration state is unstable.
+    version: Version,
+}
+
+impl SlotsTracker {
+    /// Creates a new SlotsTracker with empty slot sets and version 1.
+    pub const fn new() -> Self {
+        Self {
+            local: SlotSet::new(),
+            fully_available: SlotSet::new(),
+            partially_available: SlotSet::new(),
+            version: Version::new(),
+        }
+    }
+
+    /// Gets the current stable version.
+    ///
+    /// Returns the current version as a Version.
+    /// The internal version is always Stable.
+    pub const fn get_version(&self) -> Version {
+        // Internal invariant: self.version is always Stable
+        self.version
+    }
+
+    /// Increments the version counter using wrapping arithmetic.
+    ///
+    /// Internal invariant: This is always called on Stable versions.
+    fn increment_version(&mut self) {
+        // version is always Stable internally
+        self.version = self.version.increment();
+    }
+
+    /// Sets the local responsibility slot ranges.
+    ///
+    /// This function replaces the `local` set with the provided ranges.
+    /// It also removes the given slots from `fully_available` and `partially_available`.
+    /// If the new ranges are identical to the current local slots, no change occurs and version
+    /// is not incremented.
+    ///
+    /// # Arguments
+    ///
+    /// * `ranges` - Slice of slot ranges. Must be sorted and normalized (no overlaps, no adjacent ranges).
+    pub fn set_local_slots(&mut self, ranges: &[SlotRange]) {
+        // Check if the ranges are already equal (no change needed)
+        if self.local == ranges {
```

Comment:
Are we sure if self.local=ranges, we do not still need to remove from other sets and increment version?